File: JUNE-2025/26-06-2025/p2311.py
class Solution:
    def longestSubsequence(self, s: str, k: int) -> int:
        
        t = f"{k:b}"
        if len(t) > len(s):
            return len(s)
        
        out = s[:-len(t)].count('0')
        remove = 0
        
        for i in range(len(t)):
            c1, c2 = t[i], s[-len(t) + i]
            if c1 == '1' and c2 == '0': # s[:-len(t)] smaller than t
                break
            if c1 == '0' and c2 == '1':  # s[:-len(t) - 1]  <  t  <  s[:-len(t)] 
                remove = 1
                break
                
        return out + len(t) - remove

File: JUNE-2025/26-06-2025/p2311.cpp
class Solution {
public:
    int longestSubsequence(string s, int k) {
     int n = s.length();
    int count = 0;
    long long val = 0;
    int power = 0;
    for (int i = n - 1; i >= 0; i--) {
        if (s[i] == '0') {
            count++;
        } else {
            if (power < 32) {
                val += (1LL << power);
                if (val <= k) {
                    count++;
                } else {
                    val -= (1LL << power);
                }
            }
        }
        power++;
    }

    return count;
    }
};